lib/spack/spack/solver/concretize.lp

```diff
@@ -114,25 +114,47 @@ multiple_nodes_attribute("node_flag_source").
 multiple_nodes_attribute("depends_on").
 multiple_nodes_attribute("virtual_on_edge").
 
-% Map constraint on the literal ID to facts on the node
-attr(Name, node(min_dupe_id, A1))             :- literal(LiteralID, Name, A1), solve_literal(LiteralID).
-attr(Name, node(min_dupe_id, A1), A2)         :- literal(LiteralID, Name, A1, A2), solve_literal(LiteralID).
-attr(Name, node(min_dupe_id, A1), A2, A3)     :- literal(LiteralID, Name, A1, A2, A3), solve_literal(LiteralID), not multiple_nodes_attribute(Name).
-attr(Name, node(min_dupe_id, A1), A2, A3, A4) :- literal(LiteralID, Name, A1, A2, A3, A4), solve_literal(LiteralID).
-
-% Special cases where nodes occur in arguments other than A1
-attr("node_flag_source", node(min_dupe_id, A1), A2, node(min_dupe_id, A3)) :- literal(LiteralID, "node_flag_source", A1, A2, A3), solve_literal(LiteralID).
-attr("depends_on",       node(min_dupe_id, A1), node(min_dupe_id, A2), A3) :- literal(LiteralID, "depends_on",       A1, A2, A3), solve_literal(LiteralID).
+trigger_condition_holds(TriggerID, node(min_dupe_id, Package)) :-
+  solve_literal(TriggerID),
+  pkg_fact(Package, condition_trigger(_, TriggerID)),
+  literal(TriggerID).
+
+trigger_node(TriggerID, Node, Node) :-
+  trigger_condition_holds(TriggerID, Node),
+  literal(TriggerID).
+
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned, _, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned, _, _, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, "depends_on", _, _, Mentioned, _).
```

Comment:
These are mostly different by "arity". Wondering if we can compact these rules.

Reply:
I'm pretty sure we cannot, unless we want to make all the arguements to `attr` be tuples instead of individual elements.

Reply:
I believe this can be computed much more compactly in `asp.py`, when we emit literals. I'll give it a try and push here if I am successful.

Reply:
@becker33 I just pushed https://github.com/spack/spack/pull/40173/commits/a2b6ef84cbbfbb344e2d1cc3fe00f58e1f6e3979 which computes this fact directly from `asp.py`.